### PR TITLE
Handle nested lists of nested objects

### PR DESCRIPTION
**Pull request by mikeknep**
Fixes a bug where relational JSON could not recompose lists of nested objects. See unit test for example. I've also filed a backlog ticket to improve how we handle the edge case documented in the second unit test.

This diff is pretty noisy because I wanted to add some comments / refactors / naming improvements to help explain what's going on and reduce the amount of time spent just figuring out what's going on next time we have to visit this. The actual fix was shockingly trivial, I will call it out inline.

### CODE REVIEW

File: src/gretel_trainer/relational/json.py [hunk 1]
```diff
@@ -346,19 +364,24 @@ def _denormalize_json(
...
                 pd.DataFrame.from_dict(
                     {k: unflatten(v) for k, v in flat_dict.items()}, orient="index"
                 )
             )
-            nested_df = table_df.join(dict_df).drop(columns=col_names)
+
+            # Join the flattened dict_df onto table_df.
+            # You might think we could drop the original non-nested columns at this point,
+            # but under certain JSON shapes (e.g. nested lists of objects) we still need them present.
+            nested_df = table_df.join(dict_df)
```
Comment (mikeknep) [inline src/gretel_trainer/relational/json.py:377]: This is the only change we really had to make to fix the bug. Github is mucking up the diff a bit—compare it to the deleted line 354: we were dropping columns too aggressively, leading to `KeyError`s when performing `groupby` calls a few lines lower.

File: src/gretel_trainer/relational/json.py [hunk 2]
```diff
@@ -314,30 +314,48 @@ def _denormalize_json(
     table_name_mappings: dict[str, str],
     original_table_name: str,
 ) -> pd.DataFrame:
-    table_names = list(table_name_mappings.values())
+    # Throughout this function, "provenance name" refers to the json-breadcrumb-style name (e.g. foo^bar>baz)
```
Comment (mckornfield) [inline src/gretel_trainer/relational/json.py:317]: should this be a function docstring, or is that not appropriate since it also talks about how things are used below?

File: src/gretel_trainer/relational/json.py [hunk 3]
```diff
@@ -314,30 +314,48 @@ def _denormalize_json(
...
+    # and "node name" refers to the table name as it appears in the RelationalData graph (e.g. foo_invented_{uuid})
+
+    # The provided `table_name_mappings` argument (from producer metadata) maps from provenance name to node name.
+    # `inverse_table_name_mappings` inverts that mapping (so: node name keys, provenance name values).
+    # `table_dict` replaces the keys (node namees) in `tables` with corresponding provenance names
```
Comment (mckornfield) [inline src/gretel_trainer/relational/json.py:322]:
```suggestion
    # `table_dict` replaces the keys (node names) in `tables` with corresponding provenance names
```

File: src/gretel_trainer/relational/json.py [hunk 4]
```diff
@@ -367,16 +390,32 @@ def _denormalize_json(
                 col_name = get_parent_column_name_from_child_table_name(
                     table_provenance_name
                 )
-                table_dict[parent_name] = table_dict[parent_name].join(
```
[the remark this answers is not in the document]
Comment (mckornfield) [reply src/gretel_trainer/relational/json.py:370]: ah I like the fallback, maybe `get_table_or_empty_df` would be clearer but 🤷 maybe too wordy

File: src/gretel_trainer/relational/json.py [hunk 5]
```diff
@@ -367,16 +390,32 @@ def _denormalize_json(
...
+    table_dict: dict[str, pd.DataFrame],
+    node_name: str,
+) -> pd.DataFrame:
+    """
+    Helper function used inside _denormalize_json to safely retrieve a table dataframe.
+    If `node_name` is not present in `table_dict` (e.g. because the Gretel Job failed),
+    return an empty dataframe with the expected columns.
+    """
+    empty_fallback = pd.DataFrame(
+        data={col: [] for col in rel_data.get_table_columns(node_name)},
```
Comment (mckornfield) [inline src/gretel_trainer/relational/json.py:413]: nit, could also do `pd.DataFrame(columns=rel_data.get_table_columns(node_name))`
```
Pd.>>> pd.DataFrame(columns=["a","b","c"])
Empty DataFrame
Columns: [a, b, c]
Index: []
>>> pd.DataFrame(data={col:[] for col in ["a","b","c"]})
Empty DataFrame
Columns: [a, b, c]
Index: []
```

File: tests/relational/test_relational_data_with_json.py [hunk 6]
```diff
@@ -713,6 +713,150 @@ def test_restore_with_empty_tables(bball, invented_tables):
     assert jimmy["suspensions"] == []
 
 
+@pytest.fixture
+def nested_listed_objects(tmpdir):
+    json = """
+{
+    "Records": [
```
Comment (mckornfield) [inline tests/relational/test_relational_data_with_json.py:720]: hehe nice formatting

File: tests/relational/test_relational_data_with_json.py [hunk 7]
```diff
@@ -713,6 +713,150 @@ def test_restore_with_empty_tables(bball, invented_tables):
...
+
+    pdtest.assert_frame_equal(
+        restored["demo"],
+        pd.DataFrame(
+            data={
+                "Records": [
+                    {
+                        "userAgent": "abc",
+                        "responseElements": {
+                            "accountAttributes": [
```
Comment (mckornfield) [inline tests/relational/test_relational_data_with_json.py:843]: RIP 😢

File: tests/relational/test_relational_data_with_json.py [hunk 8]
```diff
@@ -713,6 +713,150 @@ def test_restore_with_empty_tables(bball, invented_tables):
...
+"""
+    json_df = pd.read_json(json)
+
+    rel_data = RelationalData(directory=tmpdir)
+    rel_data.add_table(name="demo", primary_key=None, data=json_df)
+
+    return rel_data
+
+
+def test_nested_listed_objects(nested_listed_objects):
```
Comment (mckornfield) [inline tests/relational/test_relational_data_with_json.py:747]: `
```suggestion
def test_nested_list_objects(nested_listed_objects):
```

File: src/gretel_trainer/relational/json.py [hunk 9]
```diff
@@ -314,30 +314,48 @@ def _denormalize_json(
...
+            fk = rel_data.get_foreign_keys(table_name)[0]
+            parent_node_name = fk.parent_table_name
+            parent_provenance_name = inverse_table_name_mappings[parent_node_name]
             col_name = get_parent_column_name_from_child_table_name(
                 table_provenance_name
             )
-            kwargs = {col_name: table_dict[parent_name].apply(lambda x: [], axis=1)}
-            table_dict[parent_name] = table_dict[parent_name].assign(**kwargs)
+            parent_df = get_table(parent_node_name)
+            kwargs = {col_name: parent_df.apply(lambda x: [], axis=1)}
```
Comment (mckornfield) [inline src/gretel_trainer/relational/json.py:346]: what does this do? makes an empty list for every row?

**Review by mckornfield — APPROVED** [review]
some questions/ nitty comments

**Review by benmccown — APPROVED** [review]
Interested in some of the answers to Matt's questions, but otherwise LGTM.

File: src/gretel_trainer/relational/json.py (hunk 9 above)
Comment (pimlock) [reply src/gretel_trainer/relational/json.py:346]: +1 to Matt's question and I think that `apply` and `assign` combo makes it a bit hard to understand.

Maybe writing it in this way would be easier to grasp?
```python
parent_df[col_name] = [[] for _ in range(len(parent_df))]
```

I think modifying `parent_df` is okay here, but if we didn't want to do that, we could
```python
parent_df.assign(**{col_name: [[] for _ in range(len(parent_df))]})
```

**Review by pimlock — APPROVED** [review]
LGTM with one suggestion for potential simplification.

**Review by gracecvking — APPROVED** [review]

File: tests/relational/test_relational_data_with_json.py (hunk 8 above)
Comment (mikeknep) [reply tests/relational/test_relational_data_with_json.py:747]: `list_objects` feels just slightly ambiguous to me, I'm going with `nested_lists_of_objects`

File: src/gretel_trainer/relational/json.py (hunk 5 above)
Comment (mikeknep) [reply src/gretel_trainer/relational/json.py:413]: Yep, yours is definitely better

File: src/gretel_trainer/relational/json.py (hunk 9 above)
Comment (mikeknep) [reply src/gretel_trainer/relational/json.py:346]: Yeah this is definitely more clear, thanks. BTW, we _need_ to modify `parent_df` so that the column we add is present on subsequent iterations through the tables loop.

File: src/gretel_trainer/relational/json.py (hunk 2 above)
Comment (mikeknep) [reply src/gretel_trainer/relational/json.py:317]: (Disclaimer: I'm not as consistent about comments and/or docstrings as I ought to be.)

This feels more like a note to developers who need to grok the implementation for debugging/modifying/refactoring, as opposed to a note to clients of the function (who won't know or need to care about "provenance" vs "node"), so to me this makes more sense as a comment than a docstring.
